kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/ColumnMapping.java

```diff
@@ -170,6 +170,194 @@ static int findMaxColumnId(StructType schema) {
     return maxColumnId;
   }
 
+  /**
+   * The following validation is performed on the updated schema
+   *
+   * <ul>
+   *   <li>Every field has column IDs and physical names, and that these are unique
+   *   <li>Array/map types have valid nested field IDs defined if icebergCompatV2 is enabled
+   *   <li>Physical column names are preserved across the update
+   * </ul>
+   *
+   * @param currentColumnPathToField column path to field for current schema
+   * @param newColumnPathToField column path to field for updated schema
+   * @param icebergCompatV2 Iceberg compatV2 is enabled
+   * @throws IllegalArgumentException in case of validation failure
+   */
+  static void validateColumnIds(
+      Map<String, StructField> currentColumnPathToField,
+      Map<String, StructField> newColumnPathToField,
+      boolean icebergCompatV2) {
+
+    Map<Long, String> currentFieldIdToPhysicalName = new HashMap<>();
+    Map<Long, List<NestedIdEntry>> currentFieldIdToNestedIds = new HashMap<>();
+
+    for (Map.Entry<String, StructField> entry : currentColumnPathToField.entrySet()) {
+      StructField field = entry.getValue();
+      long columnId = getColumnId(field);
+      currentFieldIdToPhysicalName.put(columnId, getPhysicalName(field));
+
+      // If there are nested IDs save them for subsequent consistency validation
+      if (hasNestedColumnIds(field)) {
+        currentFieldIdToNestedIds.put(columnId, nestedIdEntriesForField(field));
+      }
+    }
+
+    Map<Long, String> updatedFieldIdToPhysicalName = new HashMap<>();
+    Map<Long, List<NestedIdEntry>> updatedFieldIdToNestedIds = new HashMap<>();
+    for (Map.Entry<String, StructField> entry : newColumnPathToField.entrySet()) {
+      StructField field = entry.getValue();
+      validateFieldHasColumnIdAndPhysicalName(field);
+      long columnId = getColumnId(field);
+      // Check that it's not a duplicate column ID
+      if (updatedFieldIdToPhysicalName.containsKey(columnId)) {
+        throw new IllegalArgumentException(
+            String.format("Field %s with id %d already exists", field.getName(), columnId));
```

Comment:
I wonder if this should be a more user-facing exception? Also might be helpful to include the oldSchema and newSchema here. We can do this in a future PR though no need to block

Reply:
Same thing with the other errors we throw here, let's make debugging easier; but good to do this in a follow-up PR

Reply:
Yeah I agree, let's clean these messages up in a follow on